File: loamlab_backend/testsprite_tests/TC009_post_api_dodo_webhook_with_valid_signature.py
import requests
import hmac
import hashlib
import json
import time
import base64

import os
BASE_URL = "http://localhost:3000"
DODO_WEBHOOK_SECRET = os.getenv("DODO_WEBHOOK_SECRET", "whsec_YOUR_WEBHOOK_SECRET_PLACEHOLDER")  # 測試用 secret
TEST_USER_EMAIL = "test_dodo_user@example.com"
TIMEOUT = 30

def test_post_api_dodo_webhook_with_valid_signature():
    # 步驟 1: 獲取充值前的點數
    headers_user = {"X-User-Email": TEST_USER_EMAIL}
    params = {"email": TEST_USER_EMAIL}
    try:
        pre_resp = requests.get(f"{BASE_URL}/api/user", headers=headers_user, params=params, timeout=TIMEOUT)
        # 如果用戶不存在，預期會自動註冊 (TC004 邏輯)
        pre_points = pre_resp.json().get("points", 0) if pre_resp.status_code in (200, 201) else 0
    except Exception as e:
        pre_points = 0
        print(f"User might not exist yet: {e}")

    # 步驟 2: 構建 Dodo Webhook Payload
    timestamp = str(int(time.time()))
    msg_id = f"msg_{timestamp}"
    variant_id = "pdt_0NblmUvFrwJe36ymTELWV" # STARTER (300 pts)
    
    event_payload = {
        "type": "payment.succeeded",
        "data": {
            "payment_id": f"pay_{timestamp}",
            "customer": {
                "email": TEST_USER_EMAIL
            },
            "product_id": variant_id,
            "total_amount": 2400,
            "currency": "USD"
        }
    }
    
    payload_string = json.dumps(event_payload, separators=(',', ':'))
    
    # 步驟 3: 計算 Dodo 簽章 (Standard Webhooks format)
    # 簽署內容 = msgId.msgTimestamp.body
    signed_content = f"{msg_id}.{timestamp}.{payload_string}"
    
    # Secret 處理: 移除 whsec_ 前綴並進行 base64 解碼
    secret_key = base64.b64decode(DODO_WEBHOOK_SECRET.replace('whsec_', ''))
    
    signature = hmac.new(
        key=secret_key,
        msg=signed_content.encode("utf-8"),
        digestmod=hashlib.sha256
    ).digest()
    
    signature_b64 = base64.b64encode(signature).decode('utf-8')
    
    headers_webhook = {
        "webhook-id": msg_id,
        "webhook-timestamp": timestamp,
        "webhook-signature": f"v1,{signature_b64}",
        "Content-Type": "application/json"
    }

    # 步驟 4: 發送 Webhook
    try:
        webhook_resp = requests.post(
            f"{BASE_URL}/api/webhook", 
            headers=headers_webhook, 
            data=payload_string, 
            timeout=TIMEOUT
        )
    except Exception as e:
        raise AssertionError(f"Error posting Dodo webhook: {e}")

    assert webhook_resp.status_code == 200, f"Expected 200 status but got {webhook_resp.status_code}. Response: {webhook_resp.text}"

    # 步驟 5: 驗證點數是否增加
    try:
        post_resp = requests.get(f"{BASE_URL}/api/user", headers=headers_user, params=params, timeout=TIMEOUT)
        assert post_resp.status_code in (200, 201)
        post_points = post_resp.json().get("points", 0)
        
        # Starter 方案應增加 300 點
        assert post_points == pre_points + 300, f"Points mismatch. Before: {pre_points}, After: {post_points}, Expected gain: 300"
        print(f"✅ Dodo Webhook Verification Successful: {pre_points} -> {post_points}")
    except Exception as e:
        raise AssertionError(f"Error verifying points after webhook: {e}")

if __name__ == "__main__":
    test_post_api_dodo_webhook_with_valid_signature()

File: loamlab_backend/testsprite_tests/TC009_post_api_dodo_webhook_with_valid_signature.js
const crypto = require('crypto');
const http = require('http');

const BASE_URL = 'http://localhost:3004';
const DODO_WEBHOOK_SECRET = 'whsec_MTIzNDU2Nzg5MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTI='; 
const TEST_USER_EMAIL = 'test_dodo_user_node@example.com';

async function runTest() {
    console.log('🚀 Starting Dodo Webhook Verification (Node.js)...');

    // 1. Get pre-test points
    let prePoints = 0;
    try {
        const preData = await fetchJson(`${BASE_URL}/api/user?email=${TEST_USER_EMAIL}`, {
            'X-User-Email': TEST_USER_EMAIL
        });
        prePoints = preData.points || 0;
    } catch (e) {
        console.log('User might not exist yet, defaulting to 0 pts.');
    }

    // 2. Prepare Payload
    const timestamp = Math.floor(Date.now() / 1000).toString();
    const msgId = `msg_node_${timestamp}`;
    const variantId = 'pdt_0NblmUvFrwJe36ymTELWV'; // STARTER (300 pts)

    const payloadObj = {
        type: 'payment.succeeded',
        data: {
            payment_id: `pay_node_${timestamp}`,
            customer: { email: TEST_USER_EMAIL },
            product_id: variantId,
            total_amount: 2400,
            currency: 'USD'
        }
    };

    const payloadString = JSON.stringify(payloadObj);

    // 3. Compute Dodo Signature
    const signedContent = `${msgId}.${timestamp}.${payloadString}`;
    const secretKey = Buffer.from(DODO_WEBHOOK_SECRET.replace('whsec_', ''), 'base64');
    console.log('Secret Hex:', secretKey.toString('hex'));
    console.log('Signed Content Hex:', Buffer.from(signedContent).toString('hex'));
    
    const hmac = crypto.createHmac('sha256', secretKey);
    const digest = hmac.update(signedContent).digest('base64');
    const signatureHeader = `v1,${digest}`;

    // 4. Send Webhook
    console.log(`Sending webhook for ${TEST_USER_EMAIL}...`);
    const webhookRes = await postData(`${BASE_URL}/api/webhook`, payloadString, {
        'webhook-id': msgId,
        'webhook-timestamp': timestamp,
        'webhook-signature': signatureHeader,
        'Content-Type': 'application/json'
    });

    if (webhookRes.statusCode !== 200) {
        throw new Error(`Webhook failed: ${webhookRes.statusCode} - ${webhookRes.body}`);
    }
    console.log('✅ Webhook accepted (200 OK)');

    // 5. Verify Points
    const postDataObj = await fetchJson(`${BASE_URL}/api/user?email=${TEST_USER_EMAIL}`, {
        'X-User-Email': TEST_USER_EMAIL
    });
    const postPoints = postDataObj.points || 0;

    console.log(`📊 Points update: ${prePoints} -> ${postPoints}`);
    if (postPoints === prePoints + 300) {
        console.log('🎉 VERIFICATION SUCCESSFUL!');
    } else {
        console.error(`❌ Verification failed. Expected ${prePoints + 300}, got ${postPoints}`);
        process.exit(1);
    }
}

function fetchJson(url, headers) {
    return new Promise((resolve, reject) => {
        const u = new URL(url);
        const options = {
            hostname: u.hostname,
            port: u.port,
            path: u.pathname + u.search,
            method: 'GET',
            headers: headers
        };
        const req = http.request(options, (res) => {
            let data = '';
            res.on('data', (chunk) => data += chunk);
            res.on('end', () => {
                if (res.statusCode >= 400) reject(new Error(data));
                else resolve(JSON.parse(data));
            });
        });
        req.on('error', reject);
        req.end();
    });
}

function postData(url, body, headers) {
    return new Promise((resolve, reject) => {
        const u = new URL(url);
        const bodyBuffer = Buffer.from(body);
        const options = {
            hostname: u.hostname,
            port: u.port,
            path: u.pathname,
            method: 'POST',
            headers: {
                ...headers,
                'Content-Length': bodyBuffer.length
            }
        };
        const req = http.request(options, (res) => {
            let data = '';
            res.on('data', (chunk) => data += chunk);
            res.on('end', () => resolve({ statusCode: res.statusCode, body: data }));
        });
        req.on('error', reject);
        req.write(bodyBuffer);
        req.end();
    });
}

runTest().catch(console.error);
